libnavigation-base/src/test/java/com/mapbox/navigation/base/route/NavigationRouteWaypointsTest.kt

```diff
@@ -0,0 +1,157 @@
+package com.mapbox.navigation.base.route
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.DirectionsWaypoint
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import io.mockk.mockk
+import org.junit.Assert.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+@RunWith(Parameterized::class)
+class NavigationRouteWaypointsTest(
+    private val waypointsPerRoute: Boolean?,
+    private val responseWaypoints: List<DirectionsWaypoint>?,
+    private val routeWaypoints: List<DirectionsWaypoint>?,
+    private val expectedWaypoints: List<DirectionsWaypoint>?
+) {
+
+    companion object {
+
+        @JvmStatic
+        @Parameterized.Parameters
+        fun data(): Collection<Array<Any?>> {
+            val filledWaypoints1 = listOf(
+                DirectionsWaypoint.builder()
+                    .name("name1")
+                    .rawLocation(doubleArrayOf(1.1, 2.2))
+                    .build()
+            )
+            val filledWaypoints2 = listOf(
+                DirectionsWaypoint.builder()
+                    .name("name2")
+                    .rawLocation(doubleArrayOf(1.1, 2.2))
+                    .build()
+            )
+            return listOf(
+                // #0
+                arrayOf(null, null, null, null),
+                arrayOf(false, null, null, null),
+                arrayOf(true, null, null, null),
```

Comment:
How can it happen that a route doesn't have waypoints? Is that done for backward compatibility?

Reply:
That's the point of unit tests. Technically they can be null and the method that is being tested accounts for this. And to invoke this method we don't need to know the whole business logic of the SDK (that it can't IRL have null waypoints).

Reply:
I see your point. 

> that it can't IRL have null waypoints

should the API on `NavigationRoute` be nullable then?

Reply:
Technically they can be null. Because both `directionsRoute.waypoints()` and `directionsResponse.waypoints()` are nullable. Our backend claims one of them will be returned (depending on the waypoints_per_route parameter). 
But what if this behaviour changes? And do we really trust backend that much? You mentioned yourself that it's better to introduce custom request timeout than rely on NN.
I can return an empty list instead of course. WDYT?